File: go/eywa.go
// handler.go
package eywa

import (
    "bufio"
    "encoding/json"
    "fmt"
    "log"
    "math/rand"
    "os"
    "sync"
    "time"
)

// Structs for JSON-RPC messages
type Request struct {
    JsonRPC string      `json:"jsonrpc"`
    Method  string      `json:"method"`
    Params  interface{} `json:"params,omitempty"`
    ID      string      `json:"id,omitempty"`
}

type Response struct {
    JsonRPC string      `json:"jsonrpc"`
    Result  interface{} `json:"result,omitempty"`
    Error   interface{} `json:"error,omitempty"`
    ID      string      `json:"id,omitempty"`
}

// Global maps and mutex for concurrency
var rpcCallbacks = make(map[string]chan Response)
var handlers = make(map[string]func(Request))
var mu sync.Mutex

func init() {
    rand.Seed(time.Now().UnixNano())
}

// Register a handler for a specific method
func RegisterHandler(method string, handler func(Request)) {
    handlers[method] = handler
}

// Send a JSON-RPC request with a callback
func SendRequest(method string, params interface{}) chan Response {
    id := generateID()
    data := Request{JsonRPC: "2.0", Method: method, Params: params, ID: id}

    // Create a channel for the response and store it
    responseChan := make(chan Response, 1)
    mu.Lock()
    rpcCallbacks[id] = responseChan
    mu.Unlock()

    // Write request to stdout
    sendJSON(data)
    return responseChan
}

// Send a JSON-RPC notification (without expecting a response)
func SendNotification(method string, params interface{}) {
    data := Request{JsonRPC: "2.0", Method: method, Params: params}
    sendJSON(data)
}

// Listen for incoming data on stdin
func OpenPipe() {
    scanner := bufio.NewScanner(os.Stdin)
    for scanner.Scan() {
        var data map[string]interface{}
        if err := json.Unmarshal(scanner.Bytes(), &data); err != nil {
            log.Printf("Received invalid JSON: %v", err)
            continue
        }
        handleData(data)
    }
}

// Generate a unique ID for each request
func generateID() string {
    return fmt.Sprintf("%d", rand.Int63())
}

// Handle incoming JSON-RPC data
func handleData(data map[string]interface{}) {
    if method, ok := data["method"].(string); ok {
        handleRequest(data)
    } else if _, ok := data["id"]; ok {
        handleResponse(data)
    } else {
        log.Println("Received invalid JSON-RPC:", data)
    }
}

// Process a JSON-RPC request
func handleRequest(data map[string]interface{}) {
    method := data["method"].(string)
    request := Request{
        JsonRPC: "2.0",
        Method:  method,
        Params:  data["params"],
        ID:      data["id"].(string),
    }
    if handler, exists := handlers[method]; exists {
        handler(request)
    } else {
        log.Printf("Method %s doesn't have a registered handler", method)
    }
}

// Process a JSON-RPC response
func handleResponse(data map[string]interface{}) {
    id := data["id"].(string)
    response := Response{
        JsonRPC: "2.0",
        Result:  data["result"],
        Error:   data["error"],
        ID:      id,
    }

    mu.Lock()
    if callback, exists := rpcCallbacks[id]; exists {
        delete(rpcCallbacks, id)
        callback <- response
        close(callback)
    } else {
        log.Printf("RPC callback not registered for request with id = %s", id)
    }
    mu.Unlock()
}

// Send JSON-encoded data to stdout
func sendJSON(data interface{}) {
    encoded, err := json.Marshal(data)
    if err != nil {
        log.Printf("Failed to encode JSON: %v", err)
        return
    }
    fmt.Println(string(encoded))
}

File: py/src/eywa/eywa.py
import asyncio
import sys
import json
import os
from nanoid import generate as nanoid


rpc_callbacks = {}
handlers = {}


def handle_data(data):
    method = data.get("method")
    id_ = data.get("id")
    result = data.get("result")
    error = data.get("error")
    if method:
        handle_request(data)
    elif result and id_:
        handle_response(data)
    elif error and id_:
        handle_response(data)
    else:
        print('Received invalid JSON-RPC:\n', data)


def handle_request(data):
    method = data.get("method")
    handler = handlers.get(method)
    if handler:
        handler(data)
    else:
        print(f"Method {method} doesn't have registered handler")


def handle_response(data):
    id_ = data.get("id")
    callback = rpc_callbacks.get(id_)
    if callback is not None:
        callback.set_result(data)
        # print(f'Handling response for {callback}')
    else:
        print(f'RPC callback not registered for request with id = {id_}')


async def send_request(data):
    id_ = nanoid()
    # id_ = 10
    data["jsonrpc"] = "2.0"
    data["id"] = id_
    future = asyncio.Future()
    rpc_callbacks[id_] = future
    sys.stdout.write(json.dumps(data) + "\n")
    sys.stdout.flush()
    result = await future
    del rpc_callbacks[id_]
    return result


def send_notification(data):
    data["jsonrpc"] = "2.0"
    sys.stdout.write(json.dumps(data) + "\n")
    sys.stdout.flush()


def register_handler(method, func):
    handlers[method] = func


class LargeBufferStreamReader(asyncio.StreamReader):
    # Default limit set to 1 MB here.
    def __init__(self, limit=1024*1024*10, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._limit = limit


async def read_stdin():
    reader = LargeBufferStreamReader()
    protocol = asyncio.StreamReaderProtocol(reader)
    await asyncio.get_event_loop().connect_read_pipe(lambda: protocol, sys.stdin)

    while True:
        try:
            raw_json = await asyncio.wait_for(reader.readline(), timeout=2)
            json_data = json.loads(raw_json.decode().strip())
            handle_data(json_data)
            await asyncio.sleep(0.5)
        except asyncio.TimeoutError:
            await asyncio.sleep(0.5)


# Additional functions

SUCCESS = "SUCCESS"
ERROR = "ERROR"
PROCESSING = "PROCESSING"
EXCEPTION = "EXCEPTION"


class Sheet ():
    def __init__(self, name='Sheet'):
        self.name = name
        self.rows = []
        self.columns = []

    def add_row(self, row):
        self.rows.append(row)

    def remove_row(self, row):
        self.rows.remove(row)

    def set_columns(self, columns):
        self.columns = columns

    def toJSON(self):
        return json.dumps(self, default=lambda o:o.__dict__)


class Table ():
    def __init__(self, name='Table'):
        self.name = name
        self.sheets = []

    def add_sheet(self, sheet):
        self.sheets.append(sheet)

    def remove_sheet(self, idx=0):
        self.sheets.pop(idx)

    def toJSON(self):
        return json.dumps(self, default=lambda o:o.__dict__)


# TODO finish task reporting
class TaskReport():
    def __init__(self,message, data=None, image=None):
        self.message = message
        self.data = data
        self.image = image


# ws1 = Sheet('miroslav')
# ws1.add_row({'slaven':1,'belupo':2})
# ws1.add_row({'slaven':30,'belupo':0})


# t1 = Table('TEST')
# t1.add_sheet(ws1)

# print(t1.toJSON())
# print(json.dumps({'a':2,'b':'4444'}))


def log(event="INFO", message="", data=None, duration=None, coordinates=None, time=None):
    if time is None:
        from datetime import datetime
        time = datetime.now()

    send_notification({
        "method": "task.log",
        "params": {
            "time": time,
            "event": event,
            "message": message,
            "data": data,
            "coordinates": coordinates,
            "duration": duration
        }
    })


def info(message, data=None):
    log(event="INFO", message=message, data=data)


def error(message, data=None):
    log(event="ERROR", message=message, data=data)


def warn(message, data=None):
    log(event="WARN", message=message, data=data)


def debug(message, data=None):
    log(event="DEBUG", message=message, data=data)


def trace(message, data=None):
    log(event="TRACE", message=message, data=data)


def report(message, data=None, image=None):
    send_notification({
        'method': 'task.report',
        'params': {
            'message': message,
            'data': data,
            'image': image
        }
    })


def close_task(status="SUCCESS"):
    send_notification({
        'method': 'task.close',
        'params': {
            'status': status
        }
    })

    if status == "SUCCESS":
        exit(0)
    else:
        exit(1)


def update_task(status="PROCESSING"):
    send_notification({
        'method': 'task.update',
        'params': {
            'status': status
        }
    })


async def get_task():
    return await send_request({'method': 'task.get'})


def return_task():
    send_notification({
        'method': 'task.return'
    })
    exit(0)


async def graphql(query, variables=None):
    return await send_request({
        'method': 'eywa.datasets.graphql',
        'params': {
            'query': query,
            'variables': variables
        }
    })


__stdin__task__ = None


def open_pipe():
    global __stdin__task__
    __stdin__task__ = asyncio.create_task(read_stdin())


def exit(status=0):
    if __stdin__task__ is not None:        
        __stdin__task__.cancel()
    os.set_blocking(sys.stdin.fileno(), True)
    sys.exit(status)